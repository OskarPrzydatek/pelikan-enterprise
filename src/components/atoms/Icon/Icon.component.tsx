import { ReactComponent as CloseIcon } from '~/assets/close_icon.svg';
import { ReactComponent as HamburgerIcon } from '~/assets/hamburger_icon.svg';
import { IComponent } from '~/models';

import { IconEnum } from './Icon.enum';
import * as S from './Icon.styles';

interface IIcon extends IComponent {
  icon: IconEnum;
  width?: number;
  height?: number;
  onClick?: () => void;
}

export const Icon: React.FC<IIcon> = ({
  icon,
  width = 20,
  height,
  dataTestID,
  onClick,
}: IIcon) => {
  const renderIconByType = () => {
    switch (icon) {
      case IconEnum.HAMBURGER: {
        return (
          <HamburgerIcon
            data-testid="hamburger-icon"
            height={height}
            width={width}
          />
        );
      }
      case IconEnum.CLOSE: {
        return (
          <CloseIcon data-testid="close-icon" height={height} width={width} />
        );
      }
      default: {
        return null;
      }
    }
  };

  if (onClick) {
    return (
      <S.ClickableIcon data-testid={dataTestID} onClick={onClick}>
        {renderIconByType()}
      </S.ClickableIcon>
    );
  }

  return <>{renderIconByType()}</>;
};
