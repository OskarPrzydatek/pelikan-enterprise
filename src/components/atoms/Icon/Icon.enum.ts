export enum IconEnum {
  CHEVRON = 'CHEVRON',
  CLOSE = 'CLOSE',
  HAMBURGER = 'HAMBURGER',
}
