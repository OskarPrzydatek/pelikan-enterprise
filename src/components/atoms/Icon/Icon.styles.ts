import styled from 'styled-components';

export const ClickableIcon = styled.button`
  padding: 0;
  border: none;
  appearance: none;
  background-color: transparent;
`;
