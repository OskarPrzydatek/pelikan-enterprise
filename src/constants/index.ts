export enum Icons {
  ADD,
  ARROW,
  CHEVRON,
  CLOSE,
  DELETE,
  EDIT,
  HAMBURGER,
}

export enum Slugs {
  ATTRACTION_OVERVIEW = 'attraction-overview',
  CREATE_ATTRACTION = 'create-attraction',
  EDIT_ATTRACTION = 'edit-attraction',

  HASHTAG_OVERVIEW = 'hashtag-overview',
  CREATE_HASHTAG = 'create-hashtag',
  EDIT_HASHTAG = 'edit-hashtag',

  HOTEL_OVERVIEW = 'hotel-overview',
  CREATE_HOTEL = 'create-hotel',
  EDIT_HOTEL = 'edit-hotel',

  OFFER_OVERVIEW = 'offer-overview',
  CREATE_OFFER = 'create-offer',
  EDIT_OFFER = 'edit-offer',

  TRANSPORT_OVERVIEW = 'transport-overview',
  CREATE_TRANSPORT = 'create-transport',
  EDIT_TRANSPORT = 'edit-transport',

  USERS_OVERVIEW = 'users-overview',
  CREATE_USER = 'create-user',
  EDIT_USER = 'edit-user',

  LOGIN = 'LOGIN',

  ID = ':id',
  PAGE = ':page',
  NOT_FOUND = '*',
}

export enum Endpoints {
  ATTRACTIONS_LIST = 'attractions/list',
  ATTRACTION_BY_ID = 'attractions/get',
  CREATE_ATTRACTION = 'attractions/add',
  UPDATE_ATTRACTION = 'attractions/update',
  DELETE_ATTRACTION = 'attractions/delete',

  HASHTAGS_LIST = 'hashtags/list',
  HASHTAG_BY_ID = 'hashtags/get',
  CREATE_HASHTAG = 'hashtags/add',
  UPDATE_HASHTAG = 'hashtags/update',
  DELETE_HASHTAG = 'hashtags/delete',

  HOTELS_LIST = 'hotels/list',
  HOTEL_BY_ID = 'hotels/get',
  CREATE_HOTEL = 'hotels/add',
  UPDATE_HOTEL = 'hotels/update',
  DELETE_HOTEL = 'hotels/delete',

  OFFERS_LIST = 'offers/list',
  OFFER_BY_ID = 'offers/get',
  CREATE_OFFER = 'offers/add',
  UPDATE_OFFER = 'offers/update',
  DELETE_OFFER = 'offers/delete',

  TRANSPORTS_LIST = 'transports/list',
  TRANSPORT_BY_ID = 'transports/get',
  CREATE_TRANSPORT = 'transports/add',
  UPDATE_TRANSPORT = 'transports/update',
  DELETE_TRANSPORT = 'transports/delete',

  USERS_LIST = 'users/list',
  USER_BY_ID = 'users/get',
  CREATE_USER = 'users/add',
  UPDATE_USER = 'users/update',
  DELETE_USER = 'users/delete',
}

export const transportTypeOptions = [
  { value: 'PLANE', label: 'Samolot' },
  { value: 'BUS', label: 'Autobus' },
  { value: 'SHIP', label: 'Statek' },
];

export const userRoleType = [
  { value: 'ADMIN', label: 'Administrator' },
  { value: 'WORKER', label: 'Pracownik' },
];
