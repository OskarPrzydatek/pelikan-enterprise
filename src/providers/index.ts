export { StyledComponentsProvider } from './StyledComponents.provider';
