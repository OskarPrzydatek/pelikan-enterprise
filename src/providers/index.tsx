import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { GlobalStyleLayout } from '../styles/layouts';
import { theme } from '../styles/theme';

interface IStyledComponentsProvider {
  children: React.ReactNode;
}

interface ITestingPageProvider {
  element: JSX.Element;
}

export const StyledComponentsProvider: React.FC<IStyledComponentsProvider> = ({
  children,
}: IStyledComponentsProvider) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyleLayout />
      {children}
    </ThemeProvider>
  );
};

export const TestingPageProvider: React.FC<ITestingPageProvider> = ({
  element,
}: ITestingPageProvider) => {
  return (
    <StyledComponentsProvider>
      <MemoryRouter initialEntries={['/page1']}>
        <Routes>
          <Route element={element} path="/:page" />
        </Routes>
      </MemoryRouter>
    </StyledComponentsProvider>
  );
};
