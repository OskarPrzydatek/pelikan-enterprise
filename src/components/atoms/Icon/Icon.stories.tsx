import { Meta, StoryObj } from '@storybook/react';

import { Icons } from '~/constants';
import { StyledComponentsProvider } from '~/styles';

import { Icon } from './Icon.component';

const meta: Meta<typeof Icon> = {
  title: 'Atoms/Icon',
  component: Icon,
};

export default meta;
type Story = StoryObj<typeof Icon>;

export const ChevronIcon: Story = {
  render: () => (
    <StyledComponentsProvider>
      <Icon icon={Icons.CHEVRON} />
    </StyledComponentsProvider>
  ),
};

export const CloseIcon: Story = {
  render: () => (
    <StyledComponentsProvider>
      <Icon icon={Icons.CLOSE} />
    </StyledComponentsProvider>
  ),
};

export const HamburgerIcon: Story = {
  render: () => (
    <StyledComponentsProvider>
      <Icon icon={Icons.HAMBURGER} />
    </StyledComponentsProvider>
  ),
};
