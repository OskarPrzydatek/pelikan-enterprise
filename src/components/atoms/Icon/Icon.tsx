import { ReactComponent as ArrowIcon } from '~/assets/svg/arrow_icon.svg';
import { ReactComponent as ChevronIcon } from '~/assets/svg/chevron_icon.svg';
import { ReactComponent as CloseIcon } from '~/assets/svg/close_icon.svg';
import { ReactComponent as HamburgerIcon } from '~/assets/svg/hamburger_icon.svg';
import { Icons } from '~/constants';
import { IComponent } from '~/models';

import * as S from './Icon.styles';

interface IIcon extends IComponent {
  icon: Icons;
  width?: number;
  height?: number;
  onClick?: () => void;
}

export const Icon: React.FC<IIcon> = ({
  icon,
  width = 20,
  height,
  css,
  dataTestID,
  onClick,
}: IIcon) => {
  const renderIconByType = () => {
    switch (icon) {
      case Icons.ARROW: {
        return (
          <ArrowIcon data-testid="arrow-icon" height={height} width={width} />
        );
      }
      case Icons.CHEVRON: {
        return (
          <ChevronIcon
            data-testid="chevron-icon"
            height={height}
            width={width}
          />
        );
      }
      case Icons.CLOSE: {
        return (
          <CloseIcon data-testid="close-icon" height={height} width={width} />
        );
      }
      case Icons.HAMBURGER: {
        return (
          <HamburgerIcon
            data-testid="hamburger-icon"
            height={height}
            width={width}
          />
        );
      }
      default: {
        return null;
      }
    }
  };

  if (onClick) {
    return (
      <S.ClickableIcon css={css} data-testid={dataTestID} onClick={onClick}>
        {renderIconByType()}
      </S.ClickableIcon>
    );
  }

  return <S.IconWrapper css={css}>{renderIconByType()}</S.IconWrapper>;
};
