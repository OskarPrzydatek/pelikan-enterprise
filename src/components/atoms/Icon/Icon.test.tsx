import { fireEvent, render, screen } from '@testing-library/react';
import { describe, test, expect, vi } from 'vitest';

import { Icons } from '~/constants';
import { StyledComponentsProvider } from '~/providers';

import { Icon } from './Icon.component';

const mockOnClick = vi.fn();

describe('Icon', () => {
  test('component snapshot', () => {
    const view = render(
      <StyledComponentsProvider>
        <Icon icon={Icons.HAMBURGER} />
      </StyledComponentsProvider>
    );
    expect(view).toMatchSnapshot();
  });
  test('ensure icon show correct kind of svg icon', () => {
    const { rerender } = render(
      <StyledComponentsProvider>
        <Icon icon={'SUPER_ICON' as Icons} />
      </StyledComponentsProvider>
    );
    expect(screen.queryByTestId('super-icon')).not.toBeInTheDocument();
    rerender(
      <StyledComponentsProvider>
        <Icon icon={Icons.CHEVRON} />
      </StyledComponentsProvider>
    );
    expect(screen.getByTestId('chevron-icon')).toBeInTheDocument();
    rerender(
      <StyledComponentsProvider>
        <Icon icon={Icons.CLOSE} />
      </StyledComponentsProvider>
    );
    expect(screen.getByTestId('close-icon')).toBeInTheDocument();
    rerender(
      <StyledComponentsProvider>
        <Icon icon={Icons.HAMBURGER} />
      </StyledComponentsProvider>
    );
    expect(screen.getByTestId('hamburger-icon')).toBeInTheDocument();
  });
  test('ensure icon can be clickable', () => {
    render(
      <StyledComponentsProvider>
        <Icon dataTestID="atom-icon" icon={Icons.CLOSE} onClick={mockOnClick} />
      </StyledComponentsProvider>
    );
    fireEvent.click(screen.getByTestId('atom-icon'));
    expect(mockOnClick).toHaveBeenCalled();
  });
});
