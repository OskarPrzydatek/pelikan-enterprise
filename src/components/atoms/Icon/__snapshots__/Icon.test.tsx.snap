// Vitest Snapshot v1

exports[`Icon > component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        data-testid="hamburger-icon"
        fill="none"
        viewBox="0 0 45 23"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="21.5"
          y2="21.5"
        />
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="11.5"
          y2="11.5"
        />
        <line
          stroke="#013059"
          stroke-width="3"
          x2="45"
          y1="1.5"
          y2="1.5"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      data-testid="hamburger-icon"
      fill="none"
      viewBox="0 0 45 23"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        stroke="#013059"
        stroke-width="3"
        x2="45"
        y1="21.5"
        y2="21.5"
      />
      <line
        stroke="#013059"
        stroke-width="3"
        x2="45"
        y1="11.5"
        y2="11.5"
      />
      <line
        stroke="#013059"
        stroke-width="3"
        x2="45"
        y1="1.5"
        y2="1.5"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
